SSL and post EP

tg666 opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced Slack API integration, enabling forwarding of HTTP requests and file uploads to a configured Slack channel.
  - Added environment variable support and configuration options for NGINX, certificate management, and Slack integration.
  - Added new Makefile targets for SSL certificate generation and renewal, as well as enhanced initialization workflows.
  - Introduced NGINX service with HTTP/HTTPS support, domain redirection, gzip compression, caching, and optional access restriction.
  - Added API request handling with JSON response support on the homepage.

- **Improvements**
  - Updated Docker and service configurations for better separation of concerns and environment-specific behaviors.
  - Enhanced Makefile and service management for more robust deployment and maintenance.
  - Refined homepage handling to process API requests and deliver JSON responses with error handling.
  - Improved logging of API requests including headers and body content.

- **Bug Fixes**
  - Adjusted container names and port mappings for improved clarity and compatibility.

- **Chores**
  - Added and updated .gitignore files to exclude sensitive or generated files from version control.
  - Updated dependencies and allowed plugins in Composer configuration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tg666 committed:
SSL

- added nginx service in `docker-compose.yml`
- added Makefile commands
- added nginx configurations
- added necessary ENV variables

tg666 committed:
POST requests handling

- added dependency on package `jolicode/slack-php-api`
- added ENV variables `SLACK_TOKEN` and `SLACK_CHANNEL`
- added service of type `ApiHandlerInterface` and implementation `SlackApiHandler`
- added handling of POST requests in the `HomepagePresenter`
- removed route `/api` from `RouterFactory`

tg666 committed:
Fixed Slack message length and format issues

tg666 commented on docker/nginx/templates/nossl/web.conf.template:
```diff
@@ -0,0 +1,13 @@
+upstream app {
+    server app:8080;
+}
```
both services are under the same network so names must be "internal"

tg666 commented on src/Presenter/HomepagePresenter.php:
```diff
@@ -4,14 +4,27 @@
-        $this->getTemplate()->hoj = 'vole';
+            $this->sendResponse($response);
+        }
     }
```
implemented

tg666 commented on src/Infrastructure/ApiHandler/Slack/SlackApiHandler.php:
```diff
@@ -0,0 +1,179 @@
+                    'error' => 'Service is not available.',
+                ],
+            );
+        }
```
fixed

tg666 commented on src/Infrastructure/ApiHandler/Slack/SlackApiHandler.php:
```diff
@@ -0,0 +1,179 @@
+
+        foreach ($request->getHeaders() as $name => $value) {
+            $headers[] = "$name: $value";
+        }
```
implemented in a similar way  (+ authorization and cookie only)

tg666 commented on src/Infrastructure/ApiHandler/Slack/SlackApiHandler.php:
````diff
@@ -0,0 +1,179 @@
+                        'type' => 'section',
+                        'text' => [
+                            'type' => 'mrkdwn',
+                            'text' => '```' . $body . '```',
````
implemented

tg666 commented on docker/nginx/templates/default/web.conf.template:
```diff
@@ -0,0 +1,81 @@
+    }
+}
+
+# Handle all subdomains
+server {
+    gzip on;
+    gzip_types text/css text/plain application/javascript application/json application/xml;
+
+    set ${DOLLAR}nginx_restrict_access ${NGINX_RESTRICT_ACCESS};
+
+    listen 443 ssl;
+    listen [::]:443 ssl;
+    http2 on;
+    server_name www.${NGINX_DOMAIN_NAME};
+    ssl_certificate /etc/nginx/ssl/live/${NGINX_DOMAIN_NAME}/fullchain.pem;
```
🤔

tg666 committed:
Removed quotes from default Makefile env variables

tg666 commented on Makefile:
```diff
@@ -1,6 +1,20 @@
+NGINX_DOMAIN_NAME := "68publishers.local"
+COMPOSE_ENV := "local" # "local" or "stage"
+CERTBOT_EMAIL := "info@68publishers.io"
```
quotes removed

tg666 commented on Makefile:
```diff
@@ -12,16 +26,71 @@ restart:
+	@echo "certificates successfully created"
+
+# Stage only
+certs-renew:
+	@if [ "stage" != "${COMPOSE_ENV}" ]; then \
+  		echo "\033[1;91mError: The command certs-renew can be called in the stage environment only.\033[0m"; \
+  		exit 1; \
+  	fi
+	@docker run -it --rm --name certbot \
+		-v ./docker/nginx/certs:/etc/letsencrypt \
+		-v ./docker/certbot/www:/var/www/certbot \
+		-v ./docker/certbot/secrets:/.secrets \
+		-v ./docker/certbot/renew:/renew-hook \
+		certbot/dns-digitalocean renew \
+		--post-hook "touch /renew-hook/renewed.txt";
+	@make certs-renew.post-hook
+
```
invalid

tg666 commented on Makefile:
```diff
@@ -12,16 +26,71 @@ restart:
 	make stop
 	make start
 
+mkcert:
+	@if [ "stage" = "${COMPOSE_ENV}" ]; then \
+		docker run -it --rm --name certbot \
+			-v ./docker/nginx/certs:/etc/letsencrypt \
+			-v ./docker/certbot/www:/var/www/certbot \
+			-v ./docker/certbot/secrets:/.secrets \
+			certbot/dns-digitalocean certonly \
+			--dns-digitalocean \
+			--dns-digitalocean-credentials /.secrets/digitalocean.ini \
+			--dns-digitalocean-propagation-seconds 60 \
+			--cert-name "${NGINX_DOMAIN_NAME}" \
+			-d "${NGINX_DOMAIN_NAME}" -d "www.${NGINX_DOMAIN_NAME}" \
+			--text --agree-tos --email "${CERTBOT_EMAIL}" --rsa-key-size 4096 --verbose; \
+	else \
+		cd ./docker/nginx/certs && mkdir -p "live/${NGINX_DOMAIN_NAME}" && cd "./live/${NGINX_DOMAIN_NAME}" && mkcert -key-file privkey.pem -cert-file fullchain.pem ${NGINX_DOMAIN_NAME} "www.${NGINX_DOMAIN_NAME}"; \
+	fi
+	@echo "certificates successfully created"
+
```
already fixed

tg666 commented on src/Infrastructure/ApiHandler/Slack/SlackApiHandler.php:
```diff
@@ -0,0 +1,195 @@
+                array: $files,
+            ));
+
+            try {
+                $client->filesUploadV2(
+                    files: $files,
+                    channelId: $this->channel,
+                );
+            } catch (Throwable $e) {
+                Debugger::log($e, Debugger::ERROR);
+            }
+        }
```
we dont need it

jelen07 commented:
@tg666 There is a lot of boilerplate code in nginx. I assume this is also the case in other projects. Would it make sense to have this in a nginx image with the option to override it from the project? The nginx unit configuration behaves similarly (I quess).

jelen07 commented on src/Presenter/HomepagePresenter.php:
```diff
@@ -4,14 +4,44 @@
+                $httpResponse->setCode($httpResponse::S500_InternalServerError);
+
+                $response = new JsonResponse([
+                    'accepted' => false,
```
I would recommend to have same response structure as in other projects – to make it "more compatible", more general, less unique, when we are not planning to change it in our projects.

For example I can imagine, that could be more `erros` or other `info` messages, from AMP
```json
{
  status: "error",
  data: {
    code: 404,
    error: "Channel "started-from-the-bottom" not found."
  }
}
```

tg666 replied:
changed

tg666 commented:
> @tg666 There is a lot of boilerplate code in nginx. I assume this is also the case in other projects. Would it make sense to have this in a nginx image with the option to override it from the project? The nginx unit configuration behaves similarly (I quess).

There are similarities across projects.
However, for ngix there are only 2 configuration templates - one without SSL for the initial launch without a certificate and the other for running the application on HTTPS. However, the second configuration is not necessarily always the same (e.g. CMP does not redirect to `www`, Omni has a different configuration and other subdomains [Meili]).
Then there are folders for certbot volumes and nginx ssl certificates (certbot needs to see both).
Most of the common code is in the Makefile.

tg666 committed:
Changed POST response structure

tg666 committed:
Loggable ApiHandler

- added ApiHandler implementor `LoggableApiHandler`
- added class `Helpers` that shares request processing steps
